### PR TITLE
methods-modular_programs-hard-q7

### PR DESCRIPTION
**Pull request by ketphan02**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/methods/modular_programs/hard/q7/RevUp.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package methods.modular_programs.hard.q7;
+
+/*
+Write a Java Program with a method called revString that accepts a string and returns it in reverse. Write a main method
```
Comment (SamamthaH) [inline src/methods/modular_programs/hard/q7/RevUp.java:4]: Change to:
Write a Java Program that asks the user to enter a sentence and calls a method called revString. The method accepts the string and returns it in reverse, which the main method then prints.

File: src/methods/modular_programs/hard/q7/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import global.variables.clauses.PlaceHolder;
+import global.variables.clauses.StringLiteral;
+
+public class MainTest extends BaseTest {
+    // Java
+    public Clause[] testSentence() {
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = "Hello World";
+        return new Clause[]{
+                new StringLiteral("Enter a sentence:"),
```
Comment (SamamthaH) [inline src/methods/modular_programs/hard/q7/MainTest.java:26]: Add an optional space here

File: src/methods/modular_programs/hard/q7/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                new NewLine(),
+                new PlaceHolder(),
+        };
+    }
+
+    public void runMain() {
+        RevUp.main(new String[0]);
+    }
+
+    static Stream<Arguments> inputProvider() {
```
Comment (SamamthaH) [inline src/methods/modular_programs/hard/q7/MainTest.java:36]: Add a test case with no input and one with one character of input

File: src/methods/modular_programs/hard/q7/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void methodReturnsCorrectRev(String str, String revStr) throws Throwable {
+        Object[][] arguments = {
+                {str, String.class}
+        };
+        MethodTest m = new MethodTest(RevUp.class, "revString", arguments);
+        Object output = m.callMethod();
+        CustomAssertions._assertEquals(revStr, output, "Your revString method does not reverse the string correctly.");
```
Comment (SamamthaH) [inline src/methods/modular_programs/hard/q7/MainTest.java:52]:
```suggestion
        CustomAssertions._assertEquals(revStr, output, "Your revString method does not correctly reverse the string.");
```

File: src/methods/modular_programs/hard/q7/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void printsCorrectOutput(String str, String revStr) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct reversed string for the given input.";
+        runWithInput(str, new Clause[]{
+                new StringLiteral("Your sentence in reverse is: " + revStr)
```
Comment (SamamthaH) [inline src/methods/modular_programs/hard/q7/MainTest.java:60]: Move "Your sentence in reverse is: " up to the clause.

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — APPROVED** [review]
